FauxPose/plugin.cpp

```diff
@@ -0,0 +1,170 @@
+        return offset;
+    }
+
+    // ********************************************************************
+    virtual void set_offset(const Eigen::Quaternionf& raw_o_times_offset) override {
+        std::unique_lock   lock{offset_mutex};
+        Eigen::Quaternionf raw_o = raw_o_times_offset * offset.inverse();
+        offset                   = raw_o.inverse();
+    }
+
+    // ********************************************************************
+    virtual fast_pose_type get_fast_pose() const override {
+        // MHuzai:  In actual pose prediction, the semantics are that
+        //  we return the pose for next vsync, not now. I think we
+        //  should do the same here, unless your intent is different
+        //  with faux_pose.
+        return get_fast_pose(std::chrono::system_clock::now());
+    }
+
+    // ********************************************************************
+    // get_fast_pose(): returns a "fast_pose_type" with the algorithmically
+    //   determined location values.  (Presently moving in a circle, but
+    //   always facing "front".)
+    //
+    // NOTE: time_type == std::chrono::system_clock::time_point
+    virtual fast_pose_type get_fast_pose(time_type time) const override {
+        pose_type simulated_pose; /* The algorithmically calculated 6-DOF pose */
+        double    sim_time;       /* sim_time is used to regulate a consistent movement */
+
+        RAC_ERRNO_MSG("[fauxpose] at start of _p_one_iteration");
+
+        // Calculate simulation time from start of execution
+        std::chrono::nanoseconds elapsed_time;
+        elapsed_time = time - sim_start_time;
+        sim_time     = elapsed_time.count() * 0.000000001;
+
+        // Calculate new pose values
+        //   Pose values are calculated from the passage of time to maintain consistency */
+        simulated_pose.position[0] = amplitude * sin(sim_time * period);     // X
+        simulated_pose.position[1] = 1.5;                                    // Y
+        simulated_pose.position[2] = amplitude * cos(sim_time * period);     // Z
+        simulated_pose.orientation = Eigen::Quaternionf(1.0, 0.0, 0.0, 0.0); // (W,X,Y,Z) Facing forward
+
+        // Return the new pose
+#ifndef NDEBUG
+        std::cout << "[fauxpose] Returning pose\n";
+#endif
+        return fast_pose_type{
+            .pose = simulated_pose, .predict_computed_time = std::chrono::system_clock::now(), .predict_target_time = time};
+    }
+
+    // ---------------------------------------------------------------------
+    // ---------------------------------------------------------------------
+    // ---------------------------------------------------------------------
+private:
+    const std::shared_ptr<switchboard> sb;
+    mutable Eigen::Quaternionf         offset{Eigen::Quaternionf::Identity()};
+    mutable std::shared_mutex          offset_mutex;
+
+    time_type sim_start_time; /* Store the initial time to calculate a known runtime */
```

Comment:
`time_type` is obsolete. Please use `time_point` in `common/relative_clock.hpp`.